File: apps/langos/lib/langos/native.ex
defmodule LangOS.Native do
  @moduledoc """
  Rustler NIF bindings to LangOS Rust core.
  Falls back to pure Elixir when NIFs are unavailable.
  """
  use Rustler, otp_app: :langos, crate: "langos_nif", path: "../../crates/langos_nif"

  def tokenize(_text), do: :erlang.nif_error(:nif_not_loaded)
  def count_tokens(_text), do: :erlang.nif_error(:nif_not_loaded)
  def parse_patterns(_text, _rules_json), do: :erlang.nif_error(:nif_not_loaded)
  def build_ir(_lang, _text, _vid, _sym, _utype, _args, _conf, _engine), do: :erlang.nif_error(:nif_not_loaded)
  def detect_language(_text, _locale), do: :erlang.nif_error(:nif_not_loaded)

  @spec safe_tokenize(String.t()) :: {:ok, list()} | {:error, term()}
  def safe_tokenize(text) do
    with json when is_binary(json) <- tokenize(text),
         {:ok, tokens} <- Jason.decode(json) do
      {:ok, tokens}
    else
      _ -> fallback_tokenize(text)
    end
  rescue
    _ -> fallback_tokenize(text)
  end

  @spec safe_count_tokens(String.t()) :: non_neg_integer()
  def safe_count_tokens(text) do
    count_tokens(text)
  rescue
    _ -> text |> String.split(~r/\s+/, trim: true) |> length()
  end

  @spec safe_parse_patterns(String.t(), String.t()) :: {:ok, map() | nil}
  def safe_parse_patterns(text, rules_json) do
    case parse_patterns(text, rules_json) do
      "null" -> {:ok, nil}
      json when is_binary(json) ->
        case Jason.decode(json) do
          {:ok, nil} -> {:ok, nil}
          {:ok, map} when is_map(map) -> {:ok, map}
          _ -> {:ok, nil}
        end
    end
  rescue
    _ -> LangOS.Parser.Pattern.match(text, rules_json)
  end

  @spec safe_build_ir(keyword()) :: {:ok, map()} | {:error, term()}
  def safe_build_ir(opts) do
    language = Keyword.fetch!(opts, :language)
    text = Keyword.fetch!(opts, :text)
    vocab_id = Keyword.fetch!(opts, :vocab_id)
    symbol = Keyword.fetch!(opts, :symbol)
    unit_type = Keyword.get(opts, :unit_type, "command")
    arguments = Keyword.fetch!(opts, :arguments)
    confidence = Keyword.get(opts, :confidence, %{"overall" => 0.9, "predicate" => 0.9, "roles" => 0.9, "references" => 1.0})
    engine = Keyword.get(opts, :engine, %{"parser" => "unknown"})

    args_json = Jason.encode!(arguments)
    conf_json = Jason.encode!(confidence)
    engine_json = Jason.encode!(engine)

    with json when is_binary(json) <-
           build_ir(language, text, vocab_id, symbol, unit_type, args_json, conf_json, engine_json),
         {:ok, ir} <- Jason.decode(json) do
      {:ok, ir}
    else
      _ -> {:error, :nif_build_failed}
    end
  rescue
    _ -> {:error, :nif_not_loaded}
  end

  @spec safe_detect_language(String.t(), String.t() | nil) :: String.t()
  def safe_detect_language(text, locale_hint \\ nil) do
    detect_language(text, locale_hint)
  rescue
    _ -> locale_hint || "en"
  end

  defp fallback_tokenize(text) do
    tokens =
      Regex.scan(~r/[A-Za-z0-9']+|[^\sA-Za-z0-9]/, text)
      |> Enum.with_index()
      |> Enum.flat_map(fn
        [{token, start}, _] ->
          [%{"text" => token, "start" => start, "end" => start + String.length(token), "kind" => "word"}]
        _ -> []
      end)
    {:ok, tokens}
  end
end

File: crates/langos_nif/src/native.rs
use langos_graph::{build_ir_from_parse, ConfidenceDetail};
use langos_parser::{load_rules_from_json, parse_with_patterns};
use langos_tokenizer::{count_tokens as count_whitespace_tokens, tokenize as tokenize_text};
use rustler::{Env, NifResult, Term};
use serde_json::{json, Value};

#[rustler::nif]
fn tokenize(text: String) -> NifResult<String> {
    let tokens: Vec<Value> = tokenize_text(&text).into_iter().map(|t| {
        json!({"text": t.text, "start": t.start, "end": t.end, "kind": t.kind})
    }).collect();
    Ok(serde_json::to_string(&tokens).map_err(|e| rustler::Error::Term(Box::new(e.to_string())))?)
}

#[rustler::nif]
fn count_tokens(text: String) -> NifResult<usize> {
    Ok(count_whitespace_tokens(&text))
}

#[rustler::nif]
fn parse_patterns(text: String, rules_json: String) -> NifResult<String> {
    let rules = load_rules_from_json(&rules_json)
        .map_err(|e| rustler::Error::Term(Box::new(e)))?;

    match parse_with_patterns(&text, &rules) {
        Some(m) => {
            let out = json!({
                "rule_id": m.rule_id,
                "vocab_id": m.vocab_id,
                "symbol": m.symbol,
                "unit_type": m.unit_type,
                "arguments": m.arguments,
                "span": [m.span.0, m.span.1],
                "confidence": m.confidence
            });
            Ok(serde_json::to_string(&out).unwrap())
        }
        None => Ok("null".to_string()),
    }
}

#[rustler::nif]
fn build_ir(
    language: String,
    text: String,
    vocab_id: String,
    symbol: String,
    unit_type: String,
    arguments_json: String,
    confidence_json: String,
    engine_json: String,
) -> NifResult<String> {
    let arguments: Vec<Value> = serde_json::from_str(&arguments_json)
        .map_err(|e| rustler::Error::Term(Box::new(e.to_string())))?;

    let conf: Value = serde_json::from_str(&confidence_json)
        .unwrap_or_else(|_| json!({"overall": 0.5, "predicate": 0.5, "roles": 0.5, "references": 1.0}));
    let confidence = ConfidenceDetail {
        overall: conf["overall"].as_f64().unwrap_or(0.5),
        predicate: conf["predicate"].as_f64().unwrap_or(0.5),
        roles: conf["roles"].as_f64().unwrap_or(0.5),
        references: conf["references"].as_f64().unwrap_or(1.0),
    };

    let engine: Value = serde_json::from_str(&engine_json)
        .unwrap_or_else(|_| json!({"parser": "unknown"}));

    let ir = build_ir_from_parse(&language, &text, &vocab_id, &symbol, &unit_type, &arguments, confidence, &engine);
    Ok(serde_json::to_string(&ir).map_err(|e| rustler::Error::Term(Box::new(e.to_string())))?)
}

#[rustler::nif]
fn detect_language(text: String, locale_hint: Option<String>) -> NifResult<String> {
    Ok(detect_language_heuristic(&text, locale_hint.as_deref()))
}

fn detect_language_heuristic(text: &str, locale_hint: Option<&str>) -> String {
    if let Some(locale) = locale_hint {
        return locale.split('-').next().unwrap_or("en").to_string();
    }
    let lower = text.to_lowercase();
    if lower.contains(" muri ") || lower.starts_with("ongeramo") || lower.contains("murakoze") {
        "rw".into()
    } else if lower.contains(" ajoutez ") || lower.contains(" bonjour ") || lower.contains(" inscrire") {
        "fr".into()
    } else {
        "en".into()
    }
}

#[allow(dead_code)]
fn load(_env: Env, _info: Term) -> bool { true }
